pkg/asset/machines/aws/machines.go

```diff
@@ -0,0 +1,120 @@
+	clusterapi "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+
+	"github.com/openshift/installer/pkg/types"
+)
+
+// Machines returns a list of machines for a machinepool.
+func Machines(config *types.InstallConfig, pool *types.MachinePool, role, userDataSecret string) ([]clusterapi.Machine, error) {
+	if configPlatform := config.Platform.Name(); configPlatform != types.PlatformNameAWS {
+		return nil, fmt.Errorf("non-AWS configuration: %q", configPlatform)
+	}
+	if poolPlatform := pool.Platform.Name(); poolPlatform != types.PlatformNameAWS {
+		return nil, fmt.Errorf("non-AWS machine-pool: %q", poolPlatform)
+	}
+	clustername := config.ObjectMeta.Name
+	platform := config.Platform.AWS
+	mpool := pool.Platform.AWS
+	azs := mpool.Zones
+
+	total := int64(1)
+	if pool.Replicas != nil {
+		total = *pool.Replicas
+	}
+	numOfAZs := int64(len(azs))
+	var machines []clusterapi.Machine
+	for idx, az := range azs {
+		replicas := int32(total / numOfAZs)
+		if int64(idx) < total%numOfAZs {
+			replicas++
+		}
+
+		tags, err := tagsFromUserTags(config.ClusterID, clustername, platform.UserTags)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create awsprovider.TagSpecifications from UserTags")
+		}
+		provider := awsprovider.AWSMachineProviderConfig{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "aws.cluster.k8s.io/v1alpha1",
+				Kind:       "AWSMachineProviderConfig",
+			},
+			InstanceType:       mpool.InstanceType,
+			AMI:                awsprovider.AWSResourceReference{ID: &mpool.AMIID},
+			Tags:               tags,
+			IAMInstanceProfile: &awsprovider.AWSResourceReference{ID: pointer.StringPtr(fmt.Sprintf("%s-%s-profile", clustername, role))},
+			UserDataSecret:     &corev1.LocalObjectReference{Name: userDataSecret},
+			Subnet: awsprovider.AWSResourceReference{
+				Filters: []awsprovider.Filter{{
+					Name:   "tag:Name",
+					Values: []string{fmt.Sprintf("%s-%s-%s", clustername, role, az)},
+				}},
+			},
+			Placement: awsprovider.Placement{Region: platform.Region, AvailabilityZone: az},
+			SecurityGroups: []awsprovider.AWSResourceReference{{
+				Filters: []awsprovider.Filter{{
+					Name:   "tag:Name",
+					Values: []string{fmt.Sprintf("%s_%s_sg", clustername, role)},
+				}},
+			}},
+		}
+
+		machine := clusterapi.Machine{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "cluster.k8s.io/v1alpha1",
+				Kind:       "Machine",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "openshift-cluster-api",
+				Name:      fmt.Sprintf("%s-%s-%d", clustername, pool.Name, idx),
+				Labels: map[string]string{
+					"sigs.k8s.io/cluster-api-cluster":      clustername,
+					"sigs.k8s.io/cluster-api-machine-role": role,
+					"sigs.k8s.io/cluster-api-machine-type": role,
+				},
+			},
+			Spec: clusterapi.MachineSpec{
+				ProviderConfig: clusterapi.ProviderConfig{
+					Value: &runtime.RawExtension{Object: &provider},
+				},
+				Versions: clusterapi.MachineVersionInfo{
```

Comment:
Did you not find my [zero-value demonstration][1] convincing? ;)

[1]: https://github.com/openshift/installer/pull/573#discussion_r229904753

Reply:
these values have a meaning upstream, but set them to `""` so anybody that sees this code understands that these values to set to `""` because we have different mechanics to control them.
From the consumer end it makes no difference.

Reply:
> ... anybody that sees this code understands that these values to set to "" because we have different mechanics to control them...

This is not obvious to me from reading the code ;).  How about:

```go

Spec: clusterapi.MachineSpec{
  ProviderConfig: clusterapi.ProviderConfig{
    Value: &runtime.RawExtension{Object: &provider},
  },
  // we don't need to set Versions, because we control those via {reference our different mechanism}
}
```

Reply:
> This is not obvious to me from reading the code ;).
Well in that case, hopefully `// we don't need to set Versions, because we control those via {reference our different mechanism}` does the job

Reply:
done with https://github.com/openshift/installer/pull/573/commits/d56252219525d158af84bbd2b09d2a02b6f6ebeb -> https://github.com/openshift/installer/pull/573/commits/476be0734d68983a9335b4327c0e0e2f79dabb65